### PR TITLE
Design Wrapper

### PR DESCRIPTION
**Pull request by mcgalcode**

### CODE REVIEW

File: citrination_client/client.py [hunk 1]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
         else:
             return self._get_content_from_url(self.api_url + '/datasets/' + str(dataset_id) + '/version/' + str(version) + '/pif/' + str(uid))
 
+    def submit_experimental_design(self, data_view_id, num_candidates, target, effort, constraints=[], sampler="default"):
+
+        body = {
+            "num_candidates": num_candidates,
+            "target": target,
+            "effort": effort,
```
Comment (maxhutch) [inline citrination_client/client.py]: The effort should be less than 30.  Could you assert that and raise an error to the user if its not?

File: citrination_client/client.py [hunk 2]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
...
+            "sampler": sampler
+        }
+
+        url = self.api_url + '/data_views/' + str(data_view_id) + '/experimental_design'
+
+        print(url)
+
+        response = self._post_with_version_check(url, data=json.dumps(body), headers=self.headers).json()
+
+        print(response)
```
Comment (maxhutch) [inline citrination_client/client.py]: Could you remove this?

File: citrination_client/client.py [hunk 3]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
         else:
             return self._get_content_from_url(self.api_url + '/datasets/' + str(dataset_id) + '/version/' + str(version) + '/pif/' + str(uid))
 
+    def submit_experimental_design(self, data_view_id, num_candidates, target, effort, constraints=[], sampler="default"):
```
Comment (maxhutch) [inline citrination_client/client.py]: Default sampler should be "Default"

File: citrination_client/client.py [hunk 4]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
...
+
+        return DesignRun(response["data"]["design_run"]["uid"])
+
+    def get_experimental_design_run_status(self, data_view_id, run_uuid):
+
+        url = "{}/data_views/{}/experimental_design/{}/status".format(self.api_url, data_view_id, run_uuid)
+
+        response = self._get_with_version_check(url, headers=self.headers).json()
+
+        print(response)
```
Comment (maxhutch) [inline citrination_client/client.py]: Could you remove this?

File: citrination_client/client.py [hunk 5]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
...
+        status = response["data"]
+
+        return ProcessStatus(result=status.get(
```
Comment (maxhutch) [inline citrination_client/client.py]: The formatting here is very weird.

File: citrination_client/client.py [hunk 6]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
...
+        :param run_uuid: The UUID of the design run to kill
+        :return: The UUID of the design run
+        """
+
+        url = "{}/data_views/{}/experimental_design/{}/results".format(self.api_url, data_view_id, run_uuid)
+
+        response = self._get_with_version_check(url, headers=self.headers).json()
+
+        result = response["data"]
+        print(result)
```
Comment (maxhutch) [inline citrination_client/client.py]: Could you remove this?

File: citrination_client/client.py [hunk 7]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
...
+
+        return DesignResults(best_materials=result.get(
```
Comment (maxhutch) [inline citrination_client/client.py]: The formatting here is also weird.  What's with the line breaks?

File: citrination_client/client.py [hunk 8]
```diff
@@ -478,6 +479,81 @@ def get_pif(self, dataset_id, uid, version = None):
...
+        :param data_view_id: The ID of the data view to which the run belongs
+        :param run_uuid: The UUID of the design run to kill
+        :return: The UUID of the design run
+        """
+
+        url = "{}/data_views/{}/experimental_design/{}".format(self.api_url, data_view_id, run_uuid)
+
+        response = self._delete_with_version_check(url, headers=self.headers)
+
+        response = self._check_response_for_errors(response).json()
```
Comment (maxhutch) [inline citrination_client/client.py]: Its confusing/dangerous to overload variable names like this

File: citrination_client/client.py [hunk 9]
```diff
@@ -517,12 +593,18 @@ def _put_with_version_check(self, url, data, headers):
...
+        result = requests.delete(url, headers=headers, proxies=self.proxies)
+        return self._check_response_for_errors(result)
+
     def _check_search_response_for_errors(self, response):
         if response.status_code == 204 or response.status_code == 524:
             raise RequestTimeoutException()
         return self._check_response_for_errors(response)
 
     def _check_response_for_errors(self, response):
+        if response.status_code > 399:
```
Comment (maxhutch) [inline citrination_client/client.py]: Doesn't this supersede the `if response.status_code == 400` later in the function?

File: citrination_client/client.py (hunk 3 above)
Comment (maxhutch) [inline citrination_client/client.py]: What about the docs? What format do `target` and `constraints` need to be in, in particular?

File: citrination_client/client.py (hunk 1 above)
Comment (mcgalcode) [reply citrination_client/client.py]: 👍

File: citrination_client/client.py [hunk 10]
```diff
@@ -478,6 +479,117 @@ def get_pif(self, dataset_id, uid, version = None):
         else:
             return self._get_content_from_url(self.api_url + '/datasets/' + str(dataset_id) + '/version/' + str(version) + '/pif/' + str(uid))
 
+    def trigger_design_run(self, data_view_id, num_candidates, effort, target=None, constraints=[], sampler="Default"):
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/client.py]: 👍

File: citrination_client/client.py [hunk 11]
```diff
@@ -478,6 +479,117 @@ def get_pif(self, dataset_id, uid, version = None):
...
+        """
+        Triggers a new experimental design run
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/client.py]: 👍

File: citrination_client/client.py [hunk 12]
```diff
@@ -478,6 +479,117 @@ def get_pif(self, dataset_id, uid, version = None):
...
+
+        :param data_view_id: The ID of the data view to which the run belongs
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/client.py]: just the number 👍

File: citrination_client/client.py [hunk 13]
```diff
@@ -478,6 +479,117 @@ def get_pif(self, dataset_id, uid, version = None):
...
+            new run
+        """
+
+        if effort > 30:
+            raise CitrinationClientError("Parameter effort must be less than 30 to trigger a design run")
+
+        if target is not None:
+            target = target.to_dict()
+
+        constraint_dicts = list(
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/client.py]: 👍

File: citrination_client/client.py [hunk 14]
```diff
@@ -478,6 +479,117 @@ def get_pif(self, dataset_id, uid, version = None):
...
+
+        body = {
+            "num_candidates": num_candidates,
+            "target": target,
+            "effort": effort,
+            "constraints": constraint_dicts,
+            "sampler": sampler
+        }
+
+        url = self.api_url + '/data_views/' + str(data_view_id) + '/experimental_design'
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/client.py]: 👍

File: citrination_client/design/constraints/categorical.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
+from citrination_client.design.constraints.base import BaseConstraint
+
+class CategoricalConstraint(BaseConstraint):
+    """
+    Constrains a column to a particular set of categorical
+    values.
+    """
+
+    def __init__(self, descriptor, categories):
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/categorical.py]: 👍

File: citrination_client/client.py [hunk 16]
```diff
@@ -478,6 +479,117 @@ def get_pif(self, dataset_id, uid, version = None):
...
+
+        response = self._post_with_version_check(url, data=json.dumps(body), headers=self.headers).json()
+
+        return DesignRun(response["data"]["design_run"]["uid"])
+
+    def get_design_run_status(self, data_view_id, run_uuid):
+        """
+        Retrieves the status of an in progress design run
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/client.py]: 👍

File: citrination_client/design/constraints/categorical.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+        """
+        Constructor.
+
+        :param descriptor: The name of the column in the data
+            view to which this constraint should be applied
+        :type descriptor: str
+        :param elements: An array of categories to constrain the descriptor to
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/categorical.py]: 👍

File: citrination_client/design/constraints/elemental_composition.py [hunk 18]
```diff
@@ -0,0 +1,39 @@
+from citrination_client.design.constraints.base import BaseConstraint
+
+class ElementalCompositionConstraint(BaseConstraint):
+    """
+    Constrains a composition or inorganic formula column
+    to a particular percentage.
+    """
+
+    def __init__(self, descriptor, elements, minimum, maximum):
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/elemental_composition.py]: 👍

File: citrination_client/design/constraints/elemental_composition.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+        """
+        Constructor.
+
+        :param descriptor: The name of the column in the data
+            view to which this constraint should be applied
+        :type descriptor: str
+        :param elements: An array of element abbreviations as
+            strings, e.g. ["Mg", "C"]
+        :type elements: list of str
+        :param minimum: The minimum value (< 100) as a percentage
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/elemental_composition.py]: 👍

File: citrination_client/design/constraints/elemental_composition.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+            at which the specified elements should appear in
+            candidate compositions
+        :type minimum: float
+        :param maximum: The maximum value (< 100) as a percentage
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/elemental_composition.py]: 👍

File: citrination_client/design/constraints/elemental_composition.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+            at which the specified elements should appear in
+            candidate compositions
+        :type maximum: float
+        """
+        self._type = "elementalCompositionConstraint"
+        self._elements = elements
+        self._descriptor = descriptor
+        self._min = minimum
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/elemental_composition.py]: 👍

File: citrination_client/design/constraints/number_of_elements.py [hunk 22]
```diff
@@ -0,0 +1,30 @@
+from citrination_client.design.constraints.base import BaseConstraint
+
+class NumberOfElementsConstraint(BaseConstraint):
+    """
+    A constraint which, when applied, limits the number of elements
+    in the candidates' compositions.
+    """
+
+    def __init__(self, descriptor, minimum, maximum):
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/number_of_elements.py]: 👍

File: citrination_client/design/constraints/number_of_elements.py [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+            view to which this constraint should be applied
+        :type descriptor: str
+        :param min: The minimum number of elements in the composition
+        :type min: int
+        :param max: The maximum number of elements in the composition
+        :type max: int
+        """
+        self._type = "numberOfElementsConstraint"
+        self._descriptor = descriptor
+        self._min = minimum
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/number_of_elements.py]: rm'd this constraint

File: citrination_client/design/constraints/real_range.py [hunk 24]
```diff
@@ -0,0 +1,29 @@
+from citrination_client.design.constraints.base import BaseConstraint
+
+class RealRangeConstraint(BaseConstraint):
+    """
+    Constrains a real valued column to a range.
+    """
+
+    def __init__(self, descriptor, minimum, maximum):
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/real_range.py]: 👍

File: citrination_client/design/constraints/real_value.py [hunk 25]
```diff
@@ -0,0 +1,25 @@
+from citrination_client.design.constraints.base import BaseConstraint
+
+class RealValueConstraint(BaseConstraint):
+    """
+    Constrains a real valued column to a single value.
+    """
+
+    def __init__(self, descriptor, value=None):
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/real_value.py]: 👍

File: citrination_client/design/constraints/real_range.py [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+            view to which this constraint should be applied
+        :type descriptor: str
+        :param minimum: The minimum allowed value
+        :type minimum: float
+        :param maximum: The maximum allowed value
+        :type maximum: float
+        """
+        self._type = "real"        
+        self._descriptor = descriptor
+        self._min = minimum
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/real_range.py]: 👍

File: citrination_client/design/constraints/tests/test_constraints.py [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+
+def test_real_value_constraint():
+    """
+    Tests that a real constraint will only serialize
+    with min and max OR value but not both
+
+    Also tests that the constraint 
+    """
+
+    c_value = RealValueConstraint(descriptor="Property",value=3)
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/tests/test_constraints.py]: 👍

File: citrination_client/design/constraints/tests/test_constraints.py [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+    converted to a map with elements and logic preserved
+
+    Also tests that logic must be one of the three acceptable values
+    """
+    descriptor = "Property B"
+    elements = ["Mg", "C"]
+
+    try:
+        asdf = "asdf"
+        c = ElementalInclusionConstraint(descriptor=descriptor, elements=elements, logic=asdf)
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/tests/test_constraints.py]: 👍

File: citrination_client/design/constraints/tests/test_constraints.py [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+        should = "should"
+        ElementalInclusionConstraint(descriptor=descriptor, elements=elements, logic=should)
+        must = "must"
+        ElementalInclusionConstraint(descriptor=descriptor, elements=elements, logic=must)
+        exclude = "exclude"
+        ElementalInclusionConstraint(descriptor=descriptor, elements=elements, logic=exclude)
+        assert True
+
+    except CitrinationClientError:
+        assert False
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/constraints/tests/test_constraints.py]: 👍

File: citrination_client/design/target.py [hunk 30]
```diff
@@ -0,0 +1,32 @@
+from citrination_client.errors import CitrinationClientError
+
+class Target(object):
+    """
+    The optimization target for a design run. Consists of
+    the name of the output column to optimize and the objective
+    (either "Max" or "Min")
+    """
+
+    def __init__(self, descriptor, objective):
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/target.py]: 👍

File: citrination_client/design/tests/test_target.py [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+from citrination_client.design import Target
+
+def test_target_initialization():
+    """
+    Tests that the design target must be initialized
+    with an objective that is either Min or Max
+    """
+
+    try:
+        Target(descriptor="Band gap", objective="asdf")
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/tests/test_target.py]: 👍

File: citrination_client/design/tests/test_target.py [hunk 32]
```diff
@@ -0,0 +1,20 @@
...
+    except CitrinationClientError:
+        assert True
+
+    try:
+        Target(descriptor="Band gap", objective="Min")
+        Target(descriptor="Band gap", objective="Max")
+        assert True
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/design/tests/test_target.py]: 👍

File: citrination_client/tests/test_client.py [hunk 33]
```diff
@@ -137,3 +150,91 @@ def test_tsne(self):
...
         assert len(tsne_y["x"]) == len(tsne_y["label"]), "tSNE components x and uid had different lengths"
         assert len(tsne_y["x"]) == len(tsne_y["uid"]),   "tSNE components x and label had different lengths"
+
+    def _trigger_run(self, client, view_id, num_candidates=None, effort=None, constraints=None, target=1):
+
+        if num_candidates is None:
+            num_candidates = 10
+
+        if effort is None:
+            effort = 0
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/tests/test_client.py]: 👍

File: citrination_client/tests/test_client.py [hunk 34]
```diff
@@ -137,3 +150,91 @@ def test_tsne(self):
...
+        if constraints is None:
+            constraints = [CategoricalConstraint(descriptor="Color",
+                                                 categories=["Grey"]),
+                           ElementalInclusionConstraint(
+                                descriptor="Chemical formula",
+                                elements=["Mg"],
+                                logic="must"
+                            )]
+
+        # Use 1 as the default so that you can specify None and not have
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/tests/test_client.py]: 👍

File: citrination_client/tests/test_client.py [hunk 35]
```diff
@@ -137,3 +150,91 @@ def test_tsne(self):
...
+        client = CitrinationClient(environ["CITRINATION_API_KEY"], environ["CITRINATION_SITE"])
+        view_id = "524"
+        run = self._trigger_run(client, view_id)
+
+        try:
+            status = client.get_design_run_status(view_id, run.uuid)
+            while not status.finished():
+                time.sleep(1)
+                status = client.get_design_run_status(view_id, run.uuid)
+                print(status.progress)
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/tests/test_client.py]: 👍

File: citrination_client/tests/test_client.py [hunk 36]
```diff
@@ -137,3 +150,91 @@ def test_tsne(self):
...
+                print(status.status)
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/tests/test_client.py]: 👍

File: citrination_client/tests/test_client.py [hunk 37]
```diff
@@ -137,3 +150,91 @@ def test_tsne(self):
...
+        except Exception:
+            client.kill_design_run(view_id, run.uuid)
+            raise
```
[the remark this answers is not in the document]
Comment (mcgalcode) [reply citrination_client/tests/test_client.py]: It does re-raise.

**Pull request merged**